app/uk/gov/hmrc/disareturns/services/StreamingParserService.scala

```diff
@@ -0,0 +1,103 @@
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.Future
+import scala.util.{Failure, Success, Try}
+
+@Singleton
+class StreamingParserService @Inject() (reportingRepository: MonthlyReportingMetadataRepository, implicit val mat: Materializer) {
+
+  def validatedStream(source: Source[ByteString, _]): Source[Either[ValidationError, IsaAccount], _] =
+    source
+      .via(Framing.delimiter(ByteString("\n"), 65536, allowTruncation = false))
+      .map(_.utf8String.trim)
+      .filter(_.nonEmpty)
+      .map { line =>
+        Try(Json.parse(line)) match {
+          case Success(jsValue) =>
+            DataValidator.FirstLevelValidatorExtractNinoAndAccount(jsValue) match {
+              case Right((nino, accountNumber)) =>
+                jsValue.validate[IsaAccount] match {
+                  case JsSuccess(account, _) =>
+                    DataValidator.validate(account) match {
+                      case Right(_)  => Right(account)
+                      case Left(err) => Left(SecondLevelValidationFailure(err))
+                    }
+                  case JsError(errors) =>
+                    val domainErrors = jsErrorToDomainError(errors, nino, accountNumber)
+                    domainErrors.headOption match {
+                      case Some(error) => Left(SecondLevelValidationFailure(error))
+                      case None =>
+                        Left(
+                          SecondLevelValidationFailure(
+                            SecondLevelValidationError(
+                              nino,
+                              accountNumber,
+                              "UNKNOWN_VALIDATION",
+                              "Unknown validation error"
+                            )
+                          )
+                        )
+                    }
+                }
+              case Left(firstLevelErr) =>
+                Left(FirstLevelValidationFailure(firstLevelErr))
+            }
+          case Failure(ex) =>
+            Left(FirstLevelValidationFailure(MalformedJsonFailureErr: ErrorResponse))
+        }
+      }
+
+  def processValidatedStream(
+    isaManagerReferenceNumber: String,
+    returnId:                  String,
+    validatedStream:           Source[Either[ValidationError, IsaAccount], _]
+  ): Future[Done] =
+    validatedStream
+      .grouped(25000)
+      .mapAsync(1) { batch =>
+        val (errors, validAccounts) = batch.partitionMap(identity)
+
+        if (errors.nonEmpty) {
+          val firstLevelErrors  = errors.collect { case FirstLevelValidationFailure(err) => err }
+          val secondLevelErrors = errors.collect { case SecondLevelValidationFailure(err) => err }
+
+          val failureException = if (firstLevelErrors.nonEmpty) {
+            FirstLevelValidationException(firstLevelErrors.head)
+          } else {
+            SecondLevelValidationException(SecondLevelValidationResponse(errors = secondLevelErrors.toList))
```

Comment:
Do we not need to collect `SecondLevelVaildationError`s for the whole of the data sent to us? 

This would fail-fast after the first batch and not collect anything from the rest, if there is indeed more.

Reply:
Yep 100% correct, at the moment we rely on .grouped(25000) to collect the entire payload(This may need increasing to cover the 10mb sufficiently). This is as it stands is necessary because we want to check the entire payload to ensure each model has a account number and nino before surfacing any second level validation errors.

So yeah although its technically batched, we want to collect all the payload in one batch. Again my lack of expertise in streaming may be limiting the solution here, there may be a better way to collect everything